### PR TITLE
Role Based Access Control

### PR DESCRIPTION
**Pull request by emildoychinov**

### CODE REVIEW

File: src/ability/ability.guard.ts [hunk 1]
```diff
@@ -6,86 +6,108 @@ import { Operator } from 'src/operator/entities/operator.entity';
...
-    const permissions: Permission[] = [];
-    for(const role of operator.roles){
-      permissions.push(...(
-        await this.permissionRepository.createQueryBuilder('permissions')
-        .innerJoin('permission.roles', 'role')
-        .where('role.id = :roleID', {roleID : role.id})
-        .getMany()
-      ));
+    // if(operator.roles.some((role) => {
+    //   return role.name === SUPERUSER;
```
Comment (emil-doychinov) [inline src/ability/ability.guard.ts]: remove comment

File: src/ability/ability.guard.ts [hunk 2]
```diff
@@ -6,86 +6,108 @@ import { Operator } from 'src/operator/entities/operator.entity';
...
 
 
+
     return true;
   }
+
+  async findSubject(subject: string){
+    const table = await this.clientRepository.manager.query(
+      `SELECT EXISTS (
+        SELECT 1
```
Comment (emil-doychinov) [inline src/ability/ability.guard.ts:116]: remove unneeded function

File: src/client/client.service.ts [hunk 3]
```diff
@@ -25,12 +30,72 @@ export class ClientService {
...
-  findOne(id: number) {
-    return `This action returns a #${id} client`;
+  async findByName(name: string){
+    return this.clientRepository
+      .createQueryBuilder('client')
+      .leftJoinAndSelect('client.operators', 'operators')
+      .leftJoinAndSelect('operators.roles', 'roles')
+      .leftJoinAndSelect('operators.user', 'user')
+      .where('client.name = :name', {name})
+      .getOne();
```
Comment (emil-doychinov) [inline src/client/client.service.ts]: getOneOrFail

File: src/client/client.service.ts [hunk 4]
```diff
@@ -25,12 +30,72 @@ export class ClientService {
...
-    return `This action updates a #${id} client`;
+  async findById(id: number) {
+    return this.clientRepository
+      .createQueryBuilder('client')
+      .leftJoinAndSelect('client.operators', 'operators')
+      .leftJoinAndSelect('client.roles', 'room_roles')
+      .leftJoinAndSelect('operators.roles', 'roles')
+      .leftJoinAndSelect('operators.user', 'user')
+      .where('client.id = :id', {id})
+      .getOne();
```
Comment (emil-doychinov) [inline src/client/client.service.ts]: getOneOrFail

File: src/roles/roles.service.ts [hunk 5]
```diff
@@ -1,25 +1,127 @@
...
+    private clientService: ClientService,
+    @Inject(forwardRef(() => OperatorService))
+    private operatorService: OperatorService,
+    @Inject(PermissionsService)
+    private permissionService: PermissionsService,
+    private readonly queueService: QueueService,
+  ) {}
   create(createRoleDto: CreateRoleDto) {
-    return 'This action adds a new role';
+    //TODO : superusers can create other superusers, but regular operators cannot
```
Comment (emil-doychinov) [inline src/roles/roles.service.ts]: superusers should only be able to be created by devs? not by other superusers.

File: src/ability/ability.guard.ts [hunk 6]
```diff
@@ -6,86 +6,108 @@ import { Operator } from 'src/operator/entities/operator.entity';
...
-
-    const {bearer, token} = auth.split(' ');
-    if(bearer != 'bearer' || !token){
-      return false;
-    }
-
-    const userID = this.jwtService.decode(token).sub;
+    
+    const userID = request.user.sub;
     const operator = await this.operatorRepository.createQueryBuilder('operator')
```
Comment (emildoychinov) [inline src/ability/ability.guard.ts:58]: use service

File: src/ability/ability.guard.ts [hunk 7]
```diff
@@ -6,86 +6,108 @@ import { Operator } from 'src/operator/entities/operator.entity';
...
+        .setMessage('You are not allowed to perform this action')
+        .throwUnlessCan(rule.action, rule.subject);
     }
 
 
+
     return true;
   }
+
+  async findSubject(subject: string){
```
Comment (emildoychinov) [inline src/ability/ability.guard.ts:110]: remove unneeded function

**Pull request merged**